src/components/views/settings/tabs/user/SecurityUserSettingsTab.js

```diff
@@ -93,24 +100,62 @@ export default class SecurityUserSettingsTab extends React.Component {
         this.setState({ignoredUsers});
     };
 
-    _onRejectAllInvitesClicked = (rooms, ev) => {
+    _getInvitedRooms = () => {
+        return MatrixClientPeg.get().getRooms().filter((r) => {
+            return r.hasMembershipState(MatrixClientPeg.get().getUserId(), "invite");
+        });
+    };
+
+    _manageInvites = async (accept) => {
         this.setState({
-            rejectingInvites: true,
+            managingInvites: true,
         });
-        // reject the invites
-        const promises = rooms.map((room) => {
-            return MatrixClientPeg.get().leave(room.roomId).catch((e) => {
-                // purposefully drop errors to the floor: we'll just have a non-zero number on the UI
-                // after trying to reject all the invites.
-            });
+
+        // Compile array of invitation room ids
+        const invitedRoomIds = this._getInvitedRooms().map((room) => {
+            return room.roomId;
         });
-        Promise.all(promises).then(() => {
-            this.setState({
-                rejectingInvites: false,
+
+        // Execute all acceptances/rejections sequentially
+        const self = this;
+        const cli = MatrixClientPeg.get();
+        const action = accept ? cli.joinRoom.bind(cli) : cli.leave.bind(cli);
+        for (let i = 0; i < invitedRoomIds.length; i++) {
+            const roomId = invitedRoomIds[i];
+
+            // Accept/reject invite
+            await action(roomId).then(() => {
+                // No error, update invited rooms button
+                this.setState({invitedRoomAmt: self.state.invitedRoomAmt - 1});
+            }, async (e) => {
+                // Action failure
+                if (e.errcode === "M_LIMIT_EXCEEDED") {
+                    // Add a delay between each invite change in order to avoid rate
+                    // limiting by the server.
+                    await Promise.delay(INVITE_MANAGEMENT_DELAY);
```

Comment:
surely we should just wait for the delay specified in the error response?

Reply:
also, is this even needed? i thought that the js-sdk was smart enough to automatically back off when hitting rate limiting (which is how this used to work when originally implemented: i used it to reject 1,200 rooms from a PM spam attack on moznet one day)

Reply:
Ah, lookie that.

Reply:
@dbkr informed me that the reject all button was broken and indeed when trying to reject just ~20 rooms it would only actually reject 8 or so.

Reply:
i can believe that it regressed, but it definitely used to work (i used it!).

can you doublecheck that there isn't already ratelimiting code elsewhere in the js-sdk which should be picking this up? (i.e. are we duplicating stuff which is already there, because the existing stuff has a bug that's stopping it from kicking in properly)?  (other than this, lgtm, but would be good to check this first please)

Reply:
Here's the code for it:

https://github.com/matrix-org/matrix-js-sdk/blob/77270fa78c0615764a5b88723fc15b1390fc034b/src/client.js#L1426-L1484

I don't see anything to do with automatically retrying requests if they get an `M_LIMIT_EXCEEDED`.

The http request function:

https://github.com/matrix-org/matrix-js-sdk/blob/236397816d6f902aff680879dbc6674ce890cbf0/src/http-api.js#L675-L803

Reply:
right. so looks like js-sdk does give you a generic scheduler for handling retries, but it's only wired up for sendEvent() irritatingly. i guess this is good enough for now, but in future we really should be reusing the same scheduler & retry logic for requests of all flavours.

Reply:
Yep, totally agree.